### PR TITLE
normalize the lambdas in lambdamart objective

### PR DESCRIPTION
**Pull request by guolinke**
- [x] benchmark

benchmark run by https://github.com/guolinke/boosting_tree_benchmarks

ndcg at Yahoo LTR:

| Metric      | original | norm |
|----|  ----| ---- | 
| ndcg@1|0.72757|0.732981|
| ndcg@3|0.731267|0.735689|
| ndcg@5|0.750774|0.75352|
| ndcg@10|0.791388|0.793498|

ndcg at Yahoo LTR:

| Metric      | original | norm |
|----|  ----| ---- | 
| ndcg@1|0.512893|0.519751|
| ndcg@3|0.50118|0.499677|
| ndcg@5|0.505995|0.5058|
| ndcg@10|0.524724|0.524123|

### CODE REVIEW

File: include/LightGBM/config.h [hunk 1]
```diff
@@ -718,6 +718,10 @@ struct Config {
   // desc = optimizes `NDCG <https://en.wikipedia.org/wiki/Discounted_cumulative_gain#Normalized_DCG>`__ at this position
   int max_position = 20;
 
+  // desc = set this to ``true`` to normalize the lambdas for different queries, and improve the performance of unbalanced data
+  // desc = set this to ``false`` to enforce the original lambdamart algorithm
+  bool lambdamart_norm = false;
```
Comment (guolinke) [inline include/LightGBM/config.h]: will change to true if enable it by default is better. Will benchmark that in the next few days.

File: include/LightGBM/config.h [hunk 2]
```diff
@@ -718,6 +718,10 @@ struct Config {
   // desc = optimizes `NDCG <https://en.wikipedia.org/wiki/Discounted_cumulative_gain#Normalized_DCG>`__ at this position
   int max_position = 20;
 
+  // desc = set this to ``true`` to normalize the lambdas for different queries, and improve the performance of unbalanced data
```
Comment (StrikerRUS) [inline include/LightGBM/config.h]: The first line should be `desc = used only in ``lambdarank`` application`

File: src/objective/rank_objective.hpp [hunk 3]
```diff
@@ -104,6 +105,14 @@ class LambdarankNDCG: public ObjectiveFunction {
     }
     std::stable_sort(sorted_idx.begin(), sorted_idx.end(),
                      [score](data_size_t a, data_size_t b) { return score[a] > score[b]; });
+    // get best and worst score	
+    const double best_score = score[sorted_idx[0]];
+    data_size_t worst_idx = cnt - 1;
+    if (worst_idx > 0 && score[sorted_idx[worst_idx]] == kMinScore) {
+      worst_idx -= 1;
+    }
+    const double wrost_score = score[sorted_idx[worst_idx]];
```
Comment (StrikerRUS) [inline src/objective/rank_objective.hpp]: `worst_score`: swap `o` and `r`

**Comment by guolinke** [comment]
I think the results with norm is better, so I will enable it by default.

File: src/objective/rank_objective.hpp [hunk 4]
```diff
@@ -104,6 +105,14 @@ class LambdarankNDCG: public ObjectiveFunction {
     }
     std::stable_sort(sorted_idx.begin(), sorted_idx.end(),
                      [score](data_size_t a, data_size_t b) { return score[a] > score[b]; });
+    // get best and worst score	
+    const double best_score = score[sorted_idx[0]];
+    data_size_t wrost_idx = cnt - 1;
```
Comment (StrikerRUS) [inline src/objective/rank_objective.hpp]: Mis-correction in b84c14e2ccbf5d40c793e73625c1ebd1c232a36a: now two variables have `wrost_` prefix instead of `worst`.

**Comment by sbruch** [comment]
Is there a technical report or a paper you could share with the learning-to-rank research community that explains what this normalization does and how it was derived/it mathematically makes sense?

**Comment by guolinke** [comment]
@sbruch 
It is just a simple trick for tree-based model, not for the general learning-to-rank applications.
The idea is similar to the `reg_sqrt` in regression tasks.
In short, the tree model is sensitive to the label ranges, for its outputs are the average of label values.
For the skew distribution, the leaf outputs will be drifted to the large outputs. 
And add normalization can somehow solve this problem.

**Comment by guolinke** [comment]
@StrikerRUS is this okay to merge?

File: include/LightGBM/config.h [hunk 5]
```diff
@@ -718,6 +718,11 @@ struct Config {
   // desc = optimizes `NDCG <https://en.wikipedia.org/wiki/Discounted_cumulative_gain#Normalized_DCG>`__ at this position
   int max_position = 20;
 
+  // desc = used only in ``lambdarank`` application
+  // desc = set this to ``true`` to normalize the lambdas for different queries, and improve the performance of unbalanced data
```
Comment (StrikerRUS) [inline include/LightGBM/config.h]: `performance of` -> `performance for`

File: tests/python_package_test/test_sklearn.py [hunk 6]
```diff
@@ -79,11 +79,11 @@ def test_lambdarank(self):
...
-                eval_group=[q_test], eval_at=[1, 3], early_stopping_rounds=5, verbose=False,
-                callbacks=[lgb.reset_parameter(learning_rate=lambda x: 0.95 ** x * 0.1)])
-        self.assertLessEqual(gbm.best_iteration_, 12)
-        self.assertGreater(gbm.best_score_['valid_0']['ndcg@1'], 0.6173)
-        self.assertGreater(gbm.best_score_['valid_0']['ndcg@3'], 0.6479)
+                eval_group=[q_test], eval_at=[1, 3], early_stopping_rounds=10, verbose=False,
+                callbacks=[lgb.reset_parameter(learning_rate=lambda x: max(0.01, 0.1 - 0.01 * x))])
+        self.assertLessEqual(gbm.best_iteration_, 25)
+        self.assertGreater(gbm.best_score_['valid_0']['ndcg@1'], 0.60)
+        self.assertGreater(gbm.best_score_['valid_0']['ndcg@3'], 0.60)
```
Comment (StrikerRUS) [inline tests/python_package_test/test_sklearn.py]: Can we make these asserts more strict/precise? I remember this test helped to spot the issue of inconsistent results on different platforms due to using `sort` instead of `stable_sort`?
